### ISSUE TITLE
Collect responses issues

### ISSUE THREAD

**Issue by llizhaoxu** [MLRM-Halu/MLRM-Halu#1]
Can not directly use these scripts to generate responses. I tried to use the/
R1-Onevision with mmvp, I changed the PATH of model and data.

**Comment by Chenngzz** [comment]
Could you please provide the error message you're referring to?

**Comment by llizhaoxu** [comment]
Hi
The code of MMVP cannot run directly. The response can be well generated, but the thinking length is always 6. Here is my changed code, and the output.

`import os
import json
import argparse
import re
import torch
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from tqdm import tqdm
from transformers import AutoProcessor, AutoModelForVision2Seq
import PIL
from PIL import Image
from qwen_vl_utils import process_vision_info
np.random.seed(42)
torch.manual_seed(42)
torch.cuda.manual_seed_all(42)


def load_model_and_processor(model_path, device='cuda:2'):
    processor = AutoProcessor.from_pretrained(model_path, trust_remote_code=True)
    model = AutoModelForVision2Seq.from_pretrained(
        model_path,
        trust_remote_code=True,
        torch_dtype=torch.bfloat16 if torch.cuda.is_available() else torch.float32,
        device_map=device
    )
    model.eval()
    return model, processor


def extract_thinking_tags(text, start_tag="<think>", end_tag="</think>"):
    match = re.search(f"({re.escape(start_tag)}.*?{re.escape(end_tag)})", text, re.DOTALL)
    if match:
        thinking = match.group(1).strip()
        return thinking
    return ""


def generate_response(model, processor, image_path, question, device='cuda:2'):
    messages = [
        {
            "role": "user",
            "content": [
                {"type": "image", "image": image_path},
                {"type": "text",
                 "text": f"{question}You FIRST think about the reasoning process as an internal monologue and then provide the final answer. The reasoning process MUST BE enclosed within <think> </think> tags. The final answer MUST BE in <answer> </answer> tags."},

            ],
        }
    ]

    # Prepare inputs
    text = processor.apply_chat_template(messages, tokenize=False, add_generation_prompt=True)
    image_inputs, video_inputs = process_vision_info(messages)
    inputs = processor(
        text=[text],
        images=image_inputs,
        videos=video_inputs,
        padding=True,
        return_tensors="pt",
    )
    inputs = inputs.to(model.device)

    with torch.no_grad():
        generated_ids = model.generate(**inputs, max_new_tokens=512)
        response = processor.batch_decode(generated_ids, skip_special_tokens=True)[0]

    return response


def process_dataset(model, processor, csv_file, output_json, image_dir, num_samples=100):
    df = pd.read_csv(csv_file)[:num_samples]

    results = []
    thinking_lengths = []

    for i, row in tqdm(df.iterrows(), total=len(df)):
        image_id = str(row['Index'])
        image_path = os.path.join(image_dir, f"{image_id}.jpg")

        question = f"{row['Question']} {row['Options']}"
        answer = row['Correct Answer']
        print(image_path)
        response = generate_response(model, processor, image_path, question)
        thinking = extract_thinking_tags(response)
        thinking_token_len = len(processor.tokenizer(thinking, return_tensors="pt")["input_ids"][0]) if thinking else 0

        result = {
            "index": image_id,
            "image": image_path,
            "question": row['Question'],
            "options": row['Options'],
            "correct_answer": answer,
            "model_response": response,
            "thinking": thinking,
            "thinking_length": thinking_token_len
        }
        thinking_lengths.append(thinking_token_len)
        results.append(result)

    os.makedirs(os.path.dirname(output_json), exist_ok=True)
    with open(output_json, "w") as f:
        json.dump(results, f, indent=2)

    plt.figure(figsize=(10, 6))
    plt.hist(thinking_lengths, bins=30, edgecolor="black", alpha=0.7)
    plt.xlabel("Thinking Length (tokens)")
    plt.ylabel("Frequency")
    plt.title("Distribution of Thinking Lengths")
    plt.grid(True)
    plt.tight_layout()
    plt.savefig(os.path.join(os.path.dirname(output_json), "thinking_length_distribution.png"))

    print(f"Saved {len(results)} responses to {output_json}")



if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument('--csv', type=str, default="/home/li0007xu/Reasoning/MMVP/Questions.csv")
    parser.add_argument('--image_dir', type=str, default="/home/li0007xu/Reasoning/MMVP/Images")
    parser.add_argument('--output', type=str, default="results/openvl_output.json")
    parser.add_argument('--model_path', type=str, default="/home/li0007xu/Reasoning/R1-Onevision-7B")
    parser.add_argument('--num_samples', type=int, default=300)
    parser.add_argument('--device', type=str, default="cuda:0")
    args = parser.parse_args()

    model, processor = load_model_and_processor(args.model_path, args.device)
    process_dataset(model, processor, args.csv, args.output, args.image_dir, args.num_samples)
`



  {
    "index": "1",
    "image": "/home/li0007xu/Reasoning/MMVP/Images/1.jpg",
    "question": "Are the butterfly's wings closer to being open or closed?",
    "options": "(a) Open (b) Closed",
    "correct_answer": "(a)",
    "model_response": "system\nYou are a helpful assistant.\nuser\nAre the butterfly's wings closer to being open or closed? (a) Open (b) ClosedYou FIRST think about the reasoning process as an internal monologue and then provide the final answer. The reasoning process MUST BE enclosed within <think> </think> tags. The final answer MUST BE in <answer> </answer> tags.\nassistant\n<think>\nOkay, so I need to figure out whether the butterfly's wings are closer to being open or closed based on the image itself. Let me start by analyzing the details directly from the image.\n\nThe image shows a monarch butterfly perched on a cluster of purple flowers. The butterfly is facing left, which means its wings might be spread out since it's likely resting with them open. Monarchs typically rest with their wings open when they're not flying, which helps them conserve energy and stay warm.\n\nLooking at the wings themselves, the image clearly shows that the wings are extended outward. This indicates that the wings are open rather than folded against the body. Additionally, the image displays the intricate patterns on the wings, which would be more visible if the wings were open.\n\nThe background has green foliage, suggesting a natural setting where butterflies often rest. In such environments, butterflies usually keep their wings open to avoid predators and to regulate their body temperature. If the wings were closed, they might appear more compact or tucked under the body, but the image does not show this.\n\nFurthermore, the fact that the butterfly is perched on flowers implies it's feeding or resting. When butterflies feed, they often have their wings open to access nectar, which supports the idea that the wings are open.\n\nConsidering all these points\u2014the position of the butterfly, the extension of the wings, and the typical behavior of monarchs\u2014it seems clear that the wings are open.\n</think>\n\nThe butterfly's wings are open because the image shows them extended outward, indicating rest with wings spread, typical for monarchs in a natural setting.\n\nAnswer: (a) Open",
    "thinking": "<think> </think>",
    "thinking_length": 6
  },
